From ae3d29faac48338d3af8481f449db62bdf4d6c15 Mon Sep 17 00:00:00 2001
From: Maja Kabiljo <majakabiljo@maja-mbp.thefacebook.com>
Date: Thu, 14 Feb 2013 14:37:23 -0800
Subject: [PATCH] GIRAPH-516: out-of-core messages dies for
 ArrayIndexOutOfBoundsException when running out-of-core messages in
 UnsafeByteArrayOutputStream (majakabiljo)

---
 CHANGELOG                                                 | 2 ++
 .../giraph/comm/messages/DiskBackedMessageStore.java      | 8 +++++---
 .../main/java/org/apache/giraph/conf/GiraphConstants.java | 2 +-
 3 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index 6cc512bec..1db349258 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,8 @@
 Giraph Change Log
 
 Release 0.2.0 - unreleased
+  GIRAPH-516: out-of-core messages dies for ArrayIndexOutOfBoundsException when 
+  running out-of-core messages in UnsafeByteArrayOutputStream (majakabiljo)
 
   GIRAPH-513: OnDiskPartitionStore should take advantage of multiple disks (claudio)
 
diff --git a/giraph-core/src/main/java/org/apache/giraph/comm/messages/DiskBackedMessageStore.java b/giraph-core/src/main/java/org/apache/giraph/comm/messages/DiskBackedMessageStore.java
index 5684990d5..26abb94eb 100644
--- a/giraph-core/src/main/java/org/apache/giraph/comm/messages/DiskBackedMessageStore.java
+++ b/giraph-core/src/main/java/org/apache/giraph/comm/messages/DiskBackedMessageStore.java
@@ -114,9 +114,11 @@ boolean addVertexMessages(I vertexId,
         }
       }
 
-      for (M message : messages) {
-        message.write(extendedDataOutput);
-        numberOfMessagesInMemory.getAndIncrement();
+      synchronized (extendedDataOutput) {
+        for (M message : messages) {
+          message.write(extendedDataOutput);
+          numberOfMessagesInMemory.getAndIncrement();
+        }
       }
     } finally {
       rwLock.readLock().unlock();
diff --git a/giraph-core/src/main/java/org/apache/giraph/conf/GiraphConstants.java b/giraph-core/src/main/java/org/apache/giraph/conf/GiraphConstants.java
index 415009ce8..44d09c946 100644
--- a/giraph-core/src/main/java/org/apache/giraph/conf/GiraphConstants.java
+++ b/giraph-core/src/main/java/org/apache/giraph/conf/GiraphConstants.java
@@ -528,7 +528,7 @@ public interface GiraphConstants {
 
   /**
    * Comma-separated list of directories in the local filesystem for
-   * out-of-core partitions. 
+   * out-of-core partitions.
    */
   String PARTITIONS_DIRECTORY = "giraph.partitionsDirectory";
   /** Default directory for out-of-core partitions. */
